### DISCUSSION TITLE
User Permissions and Topic Permissions for MQTT

### DISCUSSION THREAD

**Discussion by Lithimlin** [rabbitmq/rabbitmq-server#11897] [Questions]
## Problem
I'm trying to restrict some users on my RabbitMQ, namely ones that connect via MQTT.

From what I understand from the [docs](https://www.rabbitmq.com/docs/access-control#authorisation),
the *user* (subscriber and publisher) needs permissions on the relevant vHost to:

 - Create a Queue (to subscribe)
 - Bind the MQTT topic to its queue with the relevant filters

If I want to restrict the *[topics](https://www.rabbitmq.com/docs/access-control#topic-authorisation)* the user can listen to, I have to configure that accordingly as well.

## Example Setup

 - MQTT plugin on vHost `mqtt-host` and the default `amq.topic` exchange
 - An MQTT client/user with `client-id = foobar` looking to:
   - *subscribe* to the topic `client/foobar/commands/#`
   - *subscribe* to the topic `client/fazbat/events/#`
   - *publish* to the topic `client/foobar/events/status`

## Attempt

Of course, everything works when user `foobar` has all permissions (`.* , .* , .*`) on vHost `mqtt-host` and there are no topic permissions in place.

The following *user restrictions* all lead to the subscription being denied:
```bash
rabbitmqctl set_permissions -p "mqtt-host" "foobar" "^amq\.topic$" ".*" ".*"
rabbitmqctl set_permissions -p "mqtt-host" "foobar" "^amq.*$" ".*" ".*"
rabbitmqctl set_permissions -p "mqtt-host" "foobar" ".*amq.*$" ".*" ".*"
```

Specifying just `client` in the topic permission works...:
```bash
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client\..*" "^client\..*"
```

... however, trying to specify any more than that leads to the request being denied again:
```bash
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client/.*" "^client/.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client\/.*" "^client\/.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client/{client-id}.*" "^client/{client-id}.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client\/{client-id}.*" "^client\/{client-id}.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client\.{client-id}.*" "^client\.{client-id}.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client/foobar.*" "^client/foobar.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client\/foobar.*" "^client\/foobar.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client\.foobar.*" "^client\.foobar.*"
rabbitmqctl set_topic_permissions -p "mqtt-host" "foobar" "amq.topic" "^client\.foobar.*" "^client\.fazbat.*" # read on fazbat
```

How would I correctly set up the permissions so that each user/client in my system can only do what they need to and not, for example, mess with the queues of other clients?

**Comment by michaelklishin** [comment]
Topic permissions will not help restrict access to other queues. That's what [regular permissions](https://www.rabbitmq.com/docs/access-control#authorisation) are for.

An MQTT consumer will need `queue.declare` and `basic.consume` permissions on the queue and likely `queue.bind` permissions on the exchange.

**Reply by Lithimlin** [reply]
Thanks for the pointers. I wasn't sure which ones would be needed for a consumer/subscriber.

There's two parts to this question though. How *can* I properly limit the topics a user may subscribe to?

**Reply by michaelklishin** [reply]
Using a regular expression in a topic permissions. It is matched against the topic. The same idea as with general permissions that match resource (e.g. qeueue) names.

**Reply by Lithimlin** [reply]
Yes, I know that, but as you can see above, when I try to use a regex to match the topic (with or without the `client-id` replacer), it's not matching correctly

**Reply by Lithimlin** [reply]
Also, you mention `basic.bind` for a consumer. Do you mean `queue.bind`? I can't find `basic.bind` in the docs. I assume `queue.delete` and `queue.unbind` are already covered through the other requirements?

A publisher would then need `basic.publish` on the exchange (e.g. `amq.topic` for the default exchange)?

**Reply by michaelklishin** [reply]
Yes, `queue.bind`, `basic.bind` does not exist.

**Reply by michaelklishin** [reply]
AuthN and authZ log a lot of information at debug level, including specific decisions of each backend. I don't remember if the topic permission step is logged explicitly but it'd be a very good idea to enable debug logging and then see. And start with the simplest regular expressions and topic names, e.g. something that only has one segment, e.g. `xyz`.

**Reply by michaelklishin** [reply]
Another potential reason can be the fact that even though the MQTT implementation is not based on the AMQP 0-9-1 one per se in 3.13.x and `main`, the topic routing part still is and there is a difference between the topic format between the two protocols. Specifically the topic segment separators in AMQP 0-9-1 are dots `.` and in MQTT they are slashes `/` so you might need to use the AMQP 0-9-1 format.

**Reply by Lithimlin** [reply]
## User Permissions

I'm not quite sure if I'm using `declare` with DLX or passive, but assuming it's with DLX, if I understood correctly, I need the following:

 - queue: 
   - configure (`queue.declare`)
   - write (`queue.bind`),
   - read (`queue.declare`)
 - exchange:
   - write (`queue.declare`)
   - read (`queue.bind`)
   
Assuming my exchange is `amq.topic` and my queue is `mqtt-subscription-{client-id}qos0`, the regexes should then be as follow:

- exchange: `^amq\.topic$`
- queue: `^mqtt-subscription-.*$` (for a very broad one; I'm not sure if I _can_ use `{client-id}` here.

Thus, the permission regexes would be:

- configure: `^mqtt-subscription-.*$`
- write: `^(amq\.topic|mqtt-subscription-.*)$`
- read: `^(amq\.topic|mqtt-subscription-.*)$`

Is that correct?

## Topic Permissions

Regarding the topic permissions, from the examples above it seems clear that the dot-notation is used: The write and read permissions both match anything starting with `client.`. However, if I add anything else behind that, it fails.

**Reply by Lithimlin** [reply]
Regarding the user permissions, I found that **in my case** a user needs ALL permissions (configure, write, read) for the exchange AND the topic. ~~I'm not sure why that is, but it makes sense that a `queue.declare` needs to configure the queue and `queue.bind` needs to configure the exchange, but I'm not sure why the `queue.bind` needs write access to the exchange.~~
This is because my user is both a publisher and a subscriber.

**Reply by Lithimlin** [reply]
I finally found the issue with the topic permissions: I was using `client-id` instead of `client_id`…

**Reply by Lithimlin** [reply]
> AuthN and authZ log a lot of information at debug level, including specific decisions of each backend. I don't remember if the topic permission step is logged explicitly but it'd be a very good idea to enable debug logging and then see. And start with the simplest regular expressions and topic names, e.g. something that only has one segment, e.g. `xyz`.

I don't know why I didn't think of it earlier, but checking the rabbitmq-server logs instead of my client's has been immensely helpful!

**Reply by Lithimlin** [reply]
In the topic permissions section of the docs it says
> Note that `client_id` only applies to MQTT.

Does this also apply to user permissions or only topic permissions?
